crates/matrix-sdk-ui/src/timeline/controller/mod.rs

```diff
@@ -396,25 +468,77 @@ impl<P: RoomDataProvider> TimelineController<P> {
                 Ok(has_events)
             }
 
-            TimelineFocus::Event { target: event_id, num_context_events, .. } => {
+            TimelineFocus::Event { target: event_id, num_context_events, hide_threaded_events } => {
                 let TimelineFocusKind::Event { paginator, .. } = &*self.focus else {
                     // Note: this is sync'd with code in the ctor.
                     unreachable!();
                 };
 
-                // Start a /context request, and append the results (in order) to the timeline.
-                let start_from_result = paginator
+                let event_paginator = Paginator::new(self.room_data_provider.clone());
+
+                // Start a /context request so we can know if the event is in a thread or not
+                let start_from_result = event_paginator
                     .start_from(event_id, (*num_context_events).into())
                     .await
                     .map_err(PaginationError::Paginator)?;
 
+                let thread_root_event_id = start_from_result
+                    .events
+                    .iter()
+                    .find(
+                        |event| {
+                            if let Some(id) = event.event_id() { id == *event_id } else { false }
+                        },
+                    )
+                    .and_then(|event| extract_thread_root(event.raw()));
+
+                let _ = paginator.set(match thread_root_event_id {
+                    Some(root_id) => {
+                        let tokens = event_paginator.tokens();
+                        AnyPaginator::Threaded(ThreadedEventsLoader::new(
+                            self.room_data_provider.clone(),
+                            root_id,
+                            tokens.previous,
+                            tokens.next,
+                        ))
+                    }
+                    None => AnyPaginator::Unthreaded {
+                        paginator: event_paginator,
+                        hide_threaded_events: *hide_threaded_events,
+                    },
+                });
+
                 let has_events = !start_from_result.events.is_empty();
+                let events = start_from_result.events;
 
-                self.replace_with_initial_remote_events(
-                    start_from_result.events.into_iter(),
-                    RemoteEventOrigin::Pagination,
-                )
-                .await;
+                match paginator.get() {
```

Comment:
To remove the `None` arm, can you use `paginator.get().expect("...")` as suggested in other places, please?